File: src/analogy_search.py
import argparse
import json
from pathlib import Path

from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

def load_jsonl(path: str) -> list[dict]:
    records = []
    with Path(path).open("r", encoding="utf-8") as f:
        for line in f:
            if line.strip():
                records.append(json.loads(line))
    return records

def schema_text(record: dict) -> str:
    schema = record["schema"]

    fields = [
        schema.get("story_type", ""),
        schema.get("trigger_event", ""),
        schema.get("central_conflict", ""),
        schema.get("affected_group", ""),
        schema.get("stakes", ""),
        " ".join(schema.get("causal_chain", [])),
        schema.get("narrative_schema", ""),
        schema.get("analogy_signature", ""),
    ]

    return " ".join(fields)

def find_analogies(records: list[dict], query_id: str, top_k: int = 5, prefer_cross_topic: bool = True) -> list[dict]:
    ids = [str(r["id"]) for r in records]

    if query_id not in ids:
        raise ValueError(f"Could not find article id {query_id}. Available ids: {ids}")

    query_index = ids.index(query_id)
    texts = [schema_text(r) for r in records]

    vectorizer = TfidfVectorizer(stop_words="english", ngram_range=(1, 2))
    matrix = vectorizer.fit_transform(texts)

    similarities = cosine_similarity(matrix[query_index], matrix).flatten()

    query_topic = records[query_index]["schema"].get("topic", "").lower().strip()

    results = []
    for i, score in enumerate(similarities):
        if i == query_index:
            continue

        record = records[i]
        candidate_topic = record["schema"].get("topic", "").lower().strip()

        adjusted_score = float(score)

        if prefer_cross_topic and candidate_topic == query_topic:
            adjusted_score *= 0.75

        results.append(
            {
                "id": record["id"],
                "title": record["title"],
                "topic": record["schema"].get("topic", ""),
                "story_type": record["schema"].get("story_type", ""),
                "score": round(adjusted_score, 4),
                "narrative_schema": record["schema"].get("narrative_schema", ""),
                "analogy_signature": record["schema"].get("analogy_signature", ""),
                "url": record.get("url", ""),
            }
        )

    results.sort(key=lambda x: x["score"], reverse=True)
    return results[:top_k]

def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--schemas", default="data/schema_cards.jsonl")
    parser.add_argument("--query-id", required=True)
    parser.add_argument("--top-k", type=int, default=5)
    parser.add_argument("--same-topic-ok", action="store_true")
    args = parser.parse_args()

    records = load_jsonl(args.schemas)

    query_record = next(r for r in records if str(r["id"]) == str(args.query_id))

    print("\nQUERY ARTICLE")
    print("-------------")
    print(query_record["title"])
    print(query_record["schema"]["narrative_schema"])

    results = find_analogies(
        records=records,
        query_id=str(args.query_id),
        top_k=args.top_k,
        prefer_cross_topic=not args.same_topic_ok,
    )

    print("\nSTRUCTURAL ANALOGIES")
    print("--------------------")
    for rank, result in enumerate(results, start=1):
        print(f"\n{rank}. {result['title']}")
        print(f"   Topic: {result['topic']}")
        print(f"   Story type: {result['story_type']}")
        print(f"   Score: {result['score']}")
        print(f"   Schema: {result['narrative_schema']}")
        print(f"   Analogy signature: {result['analogy_signature']}")
        if result["url"]:
            print(f"   URL: {result['url']}")

if __name__ == "__main__":
    main()

File: web/src/lib/analogySearch.ts
import type { ArticleRecord } from "./schema";

const STOPWORDS = new Set([
  "the", "a", "an", "and", "or", "but", "if", "then", "of", "to", "in", "on",
  "for", "with", "by", "from", "as", "at", "is", "are", "was", "were", "be",
  "been", "being", "this", "that", "these", "those", "it", "its", "their",
  "his", "her", "they", "them", "he", "she", "we", "you", "about", "into",
]);

function schemaText(record: ArticleRecord): string {
  const s = record.schema;

  return [
    s.story_type,
    s.trigger_event,
    s.central_conflict,
    s.affected_group,
    s.stakes,
    s.causal_chain.join(" "),
    s.narrative_schema,
    s.analogy_signature,
  ].join(" ");
}

function tokenize(text: string): string[] {
  return (
    text
      .toLowerCase()
      .match(/[a-z0-9]+/g)
      ?.filter((token) => token.length > 2 && !STOPWORDS.has(token)) ?? []
  );
}

function termCounts(tokens: string[]): Map<string, number> {
  const counts = new Map<string, number>();

  for (const token of tokens) {
    counts.set(token, (counts.get(token) ?? 0) + 1);
  }

  return counts;
}

function cosineSimilarity(a: Map<string, number>, b: Map<string, number>): number {
  let dot = 0;
  let aNorm = 0;
  let bNorm = 0;

  for (const [term, aValue] of a.entries()) {
    dot += aValue * (b.get(term) ?? 0);
    aNorm += aValue * aValue;
  }

  for (const bValue of b.values()) {
    bNorm += bValue * bValue;
  }

  if (aNorm === 0 || bNorm === 0) return 0;

  return dot / (Math.sqrt(aNorm) * Math.sqrt(bNorm));
}

export function findAnalogies(
  records: ArticleRecord[],
  queryId: string,
  topK = 5,
  preferCrossTopic = true
) {
  const queryRecord = records.find((record) => record.id === queryId);

  if (!queryRecord) {
    throw new Error(`Could not find article id ${queryId}`);
  }

  const queryVector = termCounts(tokenize(schemaText(queryRecord)));
  const queryTopic = queryRecord.schema.topic.toLowerCase().trim();

  return records
    .filter((record) => record.id !== queryId)
    .map((record) => {
      const vector = termCounts(tokenize(schemaText(record)));
      let score = cosineSimilarity(queryVector, vector);

      const candidateTopic = record.schema.topic.toLowerCase().trim();

      if (preferCrossTopic && candidateTopic === queryTopic) {
        score *= 0.75;
      }

      return {
        id: record.id,
        title: record.title,
        topic: record.schema.topic,
        story_type: record.schema.story_type,
        score,
        narrative_schema: record.schema.narrative_schema,
        analogy_signature: record.schema.analogy_signature,
        url: record.url ?? "",
      };
    })
    .sort((a, b) => b.score - a.score)
    .slice(0, topK);
}
